File: ts/money/src/exchange/converter.ts
import { ERR_INVALID_EXCHANGE_RATE, ERR_NO_CONVERTER_PROVIDED } from '#money/errors';
import type { ExchangeRates } from '#money/exchange/rates';

export class ExchangeConverter {
	public constructor(private readonly exchange: ExchangeRates) {
		if (exchange === null || exchange === undefined) {
			throw ERR_NO_CONVERTER_PROVIDED;
		}
	}

	public static create(exchange: ExchangeRates): ExchangeConverter {
		return new ExchangeConverter(exchange);
	}

	public getExchange(): ExchangeRates {
		if (this.exchange === null || this.exchange === undefined) {
			throw ERR_INVALID_EXCHANGE_RATE;
		}

		return this.exchange;
	}
}

File: go/money/exchange/converter.go
package exchange

import (
	"alloy.dev/go/money/exception"
)

// Converter provides a simple interface for currency conversion with a fixed exchange
type Converter struct {
	exchange *Rates
}

// NewConverter creates a new Converter instance with the provided Rates.
func NewConverter(exchange *Rates) (*Converter, error) {
	if exchange == nil {
		return nil, exception.ErrNoConverterProvided
	}

	return &Converter{
		exchange: exchange,
	}, nil
}

// GetExchange returns the underlying Rates instance.
func (c *Converter) GetExchange() (*Rates, error) {
	if c == nil {
		return nil, exception.ErrNoConverterProvided
	}

	if c.exchange == nil {
		return nil, exception.ErrInvalidExchangeRate
	}

	return c.exchange, nil
}
